### PR TITLE
Add full-physics emulator training and utilities

### PR DESCRIPTION
**Pull request by nbren12**
Add a machine learning model and training script which supports relative
humidity aware losses and partial fitting (to be used for online
learning in a follow-up PR). The data-structure used by this model
`ThermoBasis` may be useful for thermodynamics-aware model training in
other contexts, it allows e.g. training a model to predict relative
humidity, but with specific humidity as the loss function.

- New ThermoBasis data structure and Models/losses which take it in
- New training etnry

Added public API:
- CLI `python -m fv3fit.train_emulator`
- `fv3fit.emulation.thermobasis.ThermoBasis`
- `fv3fit.emulation.thermobasis.models.*` (see docs)
- `fv3fit.emulation.thermobasis.loss.*` (see docs)

Significant internal changes:
- Added documentation and re-organized the existing docs into separate folders.

- [x] Tests added

### CODE REVIEW

File: constraints.txt
Comment (nbren12) [inline constraints.txt:46]: MIT. approved.

File: constraints.txt
Comment (nbren12) [inline constraints.txt:56]: Apache 2 approved.

File: constraints.txt
Comment (nbren12) [inline constraints.txt:170]: MIT is approved.

File: constraints.txt
Comment (nbren12) [inline constraints.txt:185]: MIT is approved.

File: constraints.txt
Comment (nbren12) [inline constraints.txt:189]: BSD is approved.

File: constraints.txt
Comment (nbren12) [inline constraints.txt:221]: BSD is approved.

File: constraints.txt
Comment (nbren12) [inline constraints.txt:223]: BSD is approved.

File: constraints.txt
Comment (nbren12) [inline constraints.txt:239]: Python software foundation is approved (see "open source licenses" doc).

File: constraints.txt
Comment (nbren12) [inline constraints.txt:262]: MIT is approved.

**Review by nbren12 — COMMENTED** [review]
All the licenses are approved.

File: external/fv3fit/tests/emulation/thermobasis/test_thermobasis_thermo.py
Comment (nbren12) [inline external/fv3fit/tests/emulation/thermobasis/test_thermobasis_thermo.py:9]: need to refactor this to avoid hypothesis

File: external/fv3fit/fv3fit/emulation/thermobasis/loss.py
Comment (mcgibbon) [inline external/fv3fit/fv3fit/emulation/thermobasis/loss.py:95]: This docstring should probably be updated? I'm not sure what "name is None" means.

File: external/fv3fit/fv3fit/emulation/thermobasis/thermo.py
Comment (mcgibbon) [inline external/fv3fit/fv3fit/emulation/thermobasis/thermo.py:13]: Cool design, I really like having the same interfaces with properties handling conversion based on what class you're using. I have a doc suggestion and a structural suggestion.

I'd suggest updating the description of the base class to describe that it's a container of thermodynamic variables. The base class doesn't do any conversions or indicate a prognostic variable, it just says "I'm going to provide these tensor attributes".

`qc`, `u`, and `v` are a little out of place with the rest of the variables. `qc` is optional, and these three are not used in any of the thermodynamic calculations, raising the question of why they get their attributes while the scalars get a Sequence. As it's written, the ThermoBasis doesn't have a clear way to extend it with new 3D variables or to use it in a model without `u` and `v`. As a result this feels like a container that was written with a specific model in mind, instead of a general purpose class for doing conversions.

If you intend for this to be used in other models, I'd suggest separating the thermodynamic conversion logic from your model-specific choice of variables. For example, you could remove `u`, `v`, `qc`, and `scalars` from ThermoBasis and define a model-specific EmulationData class using an internal ThermoBasis with property methods for the ThermoBasis variables. Or you could use a ThermoBasis container along with another container dataclass for unrelated variables. Or you could add some kind of registration system for non-thermodynamic variables, but this would be less type-safe.

I think the core of this code is pretty applicable to other models, so it would be good to see some kind of refactor making it clear how this can be extended or used in models with no u/v/qc.

File: external/fv3fit/fv3fit/emulation/thermobasis/xarray.py
Comment (mcgibbon) [inline external/fv3fit/fv3fit/emulation/thermobasis/xarray.py:13]: Between this and Emulator, is there a Predictor class somewhere? I don't see any classes inheriting from Predictor in this PR, though they appear to follow the same APIs. it's not strictly necessary, but it can help make it clear what is going to be plugged in to a prognostic simulation.

File: external/fv3fit/fv3fit/emulation/thermobasis/emulator.py
Comment (mcgibbon) [inline external/fv3fit/fv3fit/emulation/thermobasis/emulator.py:83]: Unrelated, but would you like to do the PR refactoring the regular fv3fit training to use Hyperparameters classes for its variable listings, or should I pick that up?

File: external/fv3fit/fv3fit/emulation/thermobasis/emulator.py
Comment (mcgibbon) [inline external/fv3fit/fv3fit/emulation/thermobasis/emulator.py:199]: This doesn't look like a generic class I should subclass from if I want to make an emulation model, and nothing about it is really specific to emulation. Could you make it clear in the documentation and naming what the purpose of this is, and how a new user to this part of the code would consider making use of or extending this code (if at all)?

File: external/fv3fit/fv3fit/emulation/thermobasis/emulator.py
Comment (mcgibbon) [inline external/fv3fit/fv3fit/emulation/thermobasis/emulator.py:362]: If this is meant to be extended and there is a one to one relationship between loss targets and a model, I'd suggest having `Target.model` or even `Config.model` which allows dependency injection. Right now adding new models requires modifying this function.

File: external/fv3fit/fv3fit/emulation/thermobasis/loss.py
Comment (mcgibbon) [inline external/fv3fit/fv3fit/emulation/thermobasis/loss.py:53]: If this is for prediction at one level, could you make this clear from the name, and underscore in the docstrings whether tensors are single-level or multi-level?

File: external/fv3fit/fv3fit/emulation/thermobasis/loss.py
Comment (mcgibbon) [inline external/fv3fit/fv3fit/emulation/thermobasis/loss.py:97]: Is this saving outputs?

**Review by mcgibbon — COMMENTED** [review]
Looks great overall. I'd love to see more physically-based tensor logic in fv3fit along with good unit tests like these.

File: external/fv3fit/fv3fit/emulation/thermobasis/xarray.py
Comment (nbren12) [reply external/fv3fit/fv3fit/emulation/thermobasis/xarray.py:13]: XarrayEstimator is close to a predictor. I was planning to merge the prognostic run interfaces in a follow-up PR: https://github.com/VulcanClimateModeling/fv3net/blob/f5b2ecb3ec178e1f8524e90dd54d23cc6b49da17/workflows/prognostic_c48_run/runtime/emulator/adapter.py#L40.

I basically use

File: external/fv3fit/fv3fit/emulation/thermobasis/thermo.py
Comment (nbren12) [reply external/fv3fit/fv3fit/emulation/thermobasis/thermo.py:13]: Thanks for the thoughts. You are right that these classes are hardcoded to the emulation problem, and yes...adding qc and other variables is challenging. In return, this design makes it much easier to inject physical priors that are specific to the emulation use-case. For other problems, it's probably best to just use the functions in `fv3fit.emulation.thermo` directly.

> qc, u, and v are a little out of place with the rest of the variables.

This class contains qc, u, and v because these are all prognostic variables so need some kind of special treatment. For example, qc does not use a residual architecture.

> qc is optional

I did this to support training models that do not predict qc (which is the use-case I started development with).

File: external/fv3fit/fv3fit/emulation/thermobasis/emulator.py
Comment (nbren12) [reply external/fv3fit/fv3fit/emulation/thermobasis/emulator.py:83]: Feel free to pick it up.

File: external/fv3fit/fv3fit/emulation/thermobasis/emulator.py
Comment (nbren12) [reply external/fv3fit/fv3fit/emulation/thermobasis/emulator.py:199]: I agree this would be nice, but I'm okay violating the open-closed rule when it comes to these kinds of simple factories functions. If we start to heavily extend this model, I agree we will need a better plugin architecture. As long as this function fits on one screen, I would advocate for leaving it unchanged.

File: external/fv3fit/fv3fit/emulation/thermobasis/emulator.py
Comment (nbren12) [reply external/fv3fit/fv3fit/emulation/thermobasis/emulator.py:362]: I agree this would be nice, but I'm okay violating the open-closed rule when it comes to these kinds of simple factories functions. If we start to heavily extend this model, I agree we will need a better plugin architecture. As long as this function fits on one screen, I would advocate for leaving it unchanged.

File: external/fv3fit/fv3fit/emulation/thermobasis/emulator.py
Comment (nbren12) [reply external/fv3fit/fv3fit/emulation/thermobasis/emulator.py]: see 41c46d34baf5f15de1cb6eefc2eb8de33e806279

File: external/fv3fit/fv3fit/emulation/thermobasis/loss.py
Comment (nbren12) [reply external/fv3fit/fv3fit/emulation/thermobasis/loss.py:53]: See 1faee88d

File: external/fv3fit/fv3fit/emulation/thermobasis/loss.py
Comment (nbren12) [reply external/fv3fit/fv3fit/emulation/thermobasis/loss.py]: See 385fd85e

File: external/fv3fit/fv3fit/emulation/thermobasis/loss.py
Comment (nbren12) [reply external/fv3fit/fv3fit/emulation/thermobasis/loss.py:113]: See a831fded

File: external/fv3fit/fv3fit/emulation/thermobasis/loss.py
Comment (nbren12) [reply external/fv3fit/fv3fit/emulation/thermobasis/loss.py]: No. This just adjusted the `info` outputs of `.loss`. See a831fded.

File: external/fv3fit/fv3fit/emulation/thermobasis/thermo.py
Comment (nbren12) [reply external/fv3fit/fv3fit/emulation/thermobasis/thermo.py:13]: Ultimately, I do think we could make the data container be problem specific/hard-coded, but use Protocol in the losses to allow for static "duck-typing".

Comment (mcgibbon) [reply external/fv3fit/fv3fit/emulation/thermobasis/thermo.py:13]: OK, what I'm hearing is that this class is specific to your model (which is a particular choice of emulation problem). That's fine - my overall comment here is that it's hard to tell what's specific to your model and what's meant to be re-used by others who might be writing different emulators (e.g. an emulator of a trigger function where u/v aren't used) or non-emulator keras models (is there much of a technical fv3fit-ML distinction between emulator and non-emulator?).

This isn't purely a comment about this PR, the thermodynamic functions you referred to would also be pretty useful in any keras model, not just emulation. But it's hard to discover those re-usable parts because they're in the same namespaces as more private model-specific code (especially when that model-specific code has general-sounding names).

I get a little confused when we talk about "the emulation problem" requiring these prognostic variables, because there is not only a single emulation problem. Are you talking about FullPhysics emulation?

File: external/fv3fit/fv3fit/emulation/thermobasis/xarray.py
Comment (mcgibbon) [reply external/fv3fit/fv3fit/emulation/thermobasis/xarray.py:13]: Can you add `Predictor` as a parent class to this, or would it require significant code changes?

File: external/fv3fit/fv3fit/emulation/thermobasis/emulator.py
Comment (mcgibbon) [reply external/fv3fit/fv3fit/emulation/thermobasis/emulator.py]: Can we codify in a team meeting when we should or should not enforce OCP, and put it in our guidelines? I'm fine with this, but I've had a few PRs blocked by cases where I feel it's okay to violate the open-closed rule, and would like us to be consistent.

File: external/fv3fit/fv3fit/emulation/thermobasis/loss.py
Comment (mcgibbon) [inline external/fv3fit/fv3fit/emulation/thermobasis/loss.py:83]: Thanks for the class docstring! Can you please add a docstring to the loss function specifying whether these should have data from a single level or all levels?
```suggestion
    def loss(self, pred_rh: tf.Tensor, out: ThermoBasis) -> Tuple[Loss, Info]:
        """
        Args:
            pred_rh: predicted relative humidity at a single vertical level
            out: target values at all vertical levels
        """
```

File: external/fv3fit/fv3fit/emulation/thermobasis/thermo.py
Comment (nbren12) [reply external/fv3fit/fv3fit/emulation/thermobasis/thermo.py:13]: Are these just general comments or is there some specific request? 

Andre and I created the `fv3fit.emulation` subpackage because we couldn't think of another name, but wanted to de-couple our work from the existing `fv3fit` code. After all, this code was written without using fv3fit at all. We also considered making a completely new package, but thought it best to concentrate all tensorflow related code in fv3fit. Integrating this package better will require some more thought---e.g. the flat namespace in `fv3fit` can be laborious e.g. sometimes it's nice to use `fv3fit.emulation.online.Config` instead of ` fv3fit.OnlineEmulationConfig`. Documentation/asking a teammate seems like our standard path to improving discoverability.


> Are you talking about FullPhysics emulation?

Yes, but this could be used for non-full physics emulation too, by writing a model which simply copies the copies the input u/v to the output.

File: external/fv3fit/fv3fit/emulation/thermobasis/emulator.py
Comment (nbren12) [reply external/fv3fit/fv3fit/emulation/thermobasis/emulator.py]: I agree this is an OCP violation. Do you think it is worth addressing? I was concerned adding a registration/plugin framework would be over-engineering for this simple function. Is this function hard for you to understand? Ultimately, we could probably hook this into the fv3fit registration framework.

> I've had a few PRs blocked by cases where I feel it's okay to violate the open-closed rule, and would like us to be consistent. 

It sounds like you feel there is not consistency in our review process. The OCP is design principle, not a hard rule. Some OCP violations (e.g. changing widely-used APIs) are more potent than others. I think in a review, both the reviewer and the author have some say in whether an OCP violation is worth addressing. Sometimes a minor OCP violation (e.g. a short switch-statement in factory function) may be more complex to work around.

File: external/fv3fit/fv3fit/emulation/thermobasis/loss.py
Comment (nbren12) [reply external/fv3fit/fv3fit/emulation/thermobasis/loss.py:28]: renamed to `truth`.

File: external/fv3fit/fv3fit/emulation/thermobasis/loss.py
Comment (nbren12) [reply external/fv3fit/fv3fit/emulation/thermobasis/loss.py]: See bec9b83.

File: external/fv3fit/fv3fit/emulation/thermobasis/loss.py
Comment (nbren12) [reply external/fv3fit/fv3fit/emulation/thermobasis/loss.py]: See dbec9b83

File: external/fv3fit/fv3fit/emulation/thermobasis/xarray.py
Comment (nbren12) [reply external/fv3fit/fv3fit/emulation/thermobasis/xarray.py:13]: It's almost a one-liner, but  `Predictor.predict` requires `xr.Dataset` inputs and outputs. This gives this type error:
```
./check_types.sh
external/fv3fit/fv3fit/emulation/thermobasis/xarray.py:35: error: Argument 1 of "predict" is incompatible with supertype "Predictor"; supertype defines the argument type as "Dataset"
external/fv3fit/fv3fit/emulation/thermobasis/xarray.py:35: error: Return type "MutableMapping[Hashable, DataArray]" of "predict" incompatible with return type "Dataset" in supertype "Predictor"
```

We could probably modify this to take `Mapping[Hashable, xr.DataArray]`, but I'll leave that to another PR.

File: constraints.txt
Comment (nbren12) [inline constraints.txt:229]: apache 2 is approved.

File: constraints.txt
Comment (nbren12) [inline constraints.txt:102]: - [ ] Mozilla Public license v2 is approved for apache 2, but need to confirm approval for MIT

File: external/fv3fit/fv3fit/emulation/thermobasis/emulator.py
Comment (mcgibbon) [reply external/fv3fit/fv3fit/emulation/thermobasis/emulator.py]: > I agree this is an OCP violation. Do you think it is worth addressing?

I said I am fine with this. If this is meant to be re-used by other people for other models I'd ask for it to be addressed, but my understanding is this is to be used in this one model.

I don't think there is uniform understanding across the team about what the OCP means, and it would help to have a team discussion about what it is and when we should or shouldn't follow it. It's a good principle that the team should develop a better understanding of. Each of the [six times](https://github.com/VulcanClimateModeling/fv3net/pulls?q=%22open-closed+principle%22+in%3Acomments) the open/closed principle it has been mentioned by a reviewer, you were the reviewer. So yes, it does feel like there is inconsistency in the review process.

File: external/fv3fit/fv3fit/emulation/thermobasis/thermo.py
Comment (mcgibbon) [reply external/fv3fit/fv3fit/emulation/thermobasis/thermo.py:13]: > Are these just general comments or is there some specific request?

> If you have any ideas on how to improve this I'd appreciate it, though there's no specific PR-blocking change I'm asking for here.

OK, it sounds like we have three development concepts at the same level then - sklearn, keras, and emulation. emulation is like an alternative of keras which was done for emulation. What I'd like to avoid is having `keras` and `emulation` depend on each other at this level, without completely integrating them. I think moving forward if we use code from `emulation` in a model under `keras`, it would make sense to first move that code to `_shared`.

Everything looks good on this for this PR.

**Review by mcgibbon — APPROVED** [review]
LGTM! Thanks for adding all the information I asked for.

File: external/fv3fit/fv3fit/emulation/thermobasis/emulator.py
Comment (mcgibbon) [reply external/fv3fit/fv3fit/emulation/thermobasis/emulator.py]: To be clear, I'm not saying there's anything wrong with your reviews, I just think this is a particularly relevant case where a principle is kind of hard to understand, and not everyone in the team is taking it into account in their reviews.

**Comment by nbren12** [comment]
I'm going to wait until @frodre has reviewed and we hear back about the license for hypothesis.

File: external/fv3fit/fv3fit/emulation/thermobasis/thermo.py
Comment (nbren12) [reply external/fv3fit/fv3fit/emulation/thermobasis/thermo.py:13]: > I think moving forward if we use code from emulation in a model under keras, it would make sense to first move that code to _shared.

Agreed.

File: constraints.txt
Comment (nbren12) [reply constraints.txt:102]: MPL v2 is compatible with MIT. Let's merge this PR and then relicense to Apache 2.

File: external/fv3fit/fv3fit/emulation/thermobasis/emulator.py
Comment (frodre) [inline external/fv3fit/fv3fit/emulation/thermobasis/emulator.py:280]: Remove if unnecessary.

File: external/fv3fit/fv3fit/emulation/thermobasis/emulator.py
Comment (frodre) [inline external/fv3fit/fv3fit/emulation/thermobasis/emulator.py:290]: Feels a bit cumbersome to have all the tf.ds batch -> basis handling within the model.  Limitations of tf dataset usage.  Maybe some sort of `BasisIterator` could be useful down the line?

File: external/fv3fit/fv3fit/emulation/thermobasis/xarray.py
Comment (frodre) [inline external/fv3fit/fv3fit/emulation/thermobasis/xarray.py:70]: Some of these might be useful in the `emulation.data.transform` module when this module is refactored.

File: external/fv3fit/tests/emulation/thermobasis/test_emulator.py
Comment (frodre) [inline external/fv3fit/tests/emulation/thermobasis/test_emulator.py:39]: Do the logs need to be checked?  I'm guessing just having the writer available triggers the TBLogger

**Review by frodre — APPROVED** [review]
Whew!  Thanks for your patience.  A lot of nice additions here!  In general, looks good to me, but there are a few minor suggestions and questions.

File: external/fv3fit/tests/emulation/thermobasis/test_models.py
Comment (nbren12) [reply external/fv3fit/tests/emulation/thermobasis/test_models.py:56]: Yah. Just figuring out how tensorflow works. The clean code books recommends turning these kinds of experiments into tests rather than just leaving them as ipython exploration.

File: external/fv3fit/tests/emulation/thermobasis/test_models.py
Comment (nbren12) [reply external/fv3fit/tests/emulation/thermobasis/test_models.py:69]: I change the name of the function to `test_ScalarMLP_integration_with_loss` and expanded the comment.

File: external/fv3fit/tests/emulation/thermobasis/test_models.py
Comment (nbren12) [reply external/fv3fit/tests/emulation/thermobasis/test_models.py:1]: I tried to do this by embellishing the test function names and adding comments/docstrings.

File: external/fv3fit/fv3fit/emulation/thermobasis/emulator.py
Comment (nbren12) [reply external/fv3fit/fv3fit/emulation/thermobasis/emulator.py:290]: I wouldn't say `Trainer` is the "model" it is the training loop. I see it is an intermediary between the dataset and the actual ML models, so this data transformations seems appropriate. I started this code by pushing this translation logic mostly into the data pipeline, but found this tightly coupled the code in `fv3fit.emulation.data` to the particular ML problem. E.g adding a new input variable required changes to both `fv3fit.emulation.data` etc.

File: external/fv3fit/tests/emulation/thermobasis/test_emulator.py
Comment (nbren12) [reply external/fv3fit/tests/emulation/thermobasis/test_emulator.py:39]: It's a bit hard to check automatically, but the code is triggered.

**Pull request merged**
